DietPi-Sync | Resolve rsync option string bug

MichaIng opened the pull request:
**Status**: Ready
- [x] Changelog
- [x] Fix inc_excl paths, which need to be relative to source dir

**Testing**:
- [x] VM Jessie
- [x] VM Stretch
- [x] VM Buster

**Reference**: https://dietpi.com/phpbb/viewtopic.php?f=11&t=5347

**Commit list/description**:
+ DietPi-Sync | Resolved an issue with wrong created rsync_options string
+ DietPi-Cron.daily | Fix daily dietpi-sync based on new settings file scheme
+ DietPi-Sync | Resolved doubled dry run due to dry run required for space check
+ DietPi-Sync | Removed compression option, since this is only relevant for remote transfer, which we do not support yet. This might have led to wrong end user assumptions and confusion.
+ DietPi-Sync | Only run real sync, if prior dry run was successful
+ DietPi-Sync | Show now last log entry instead of last sync date only in main menu
+ DietPi-Sync | Simplify logging: Log to $FP_TARGET only, create tmp log first, move to $FP_TARGET afterwards; Return after insufficient space, to skip obsolete final whip and log file view.
+ DietPi-Sync | Be more verbose when target dir pre-creation fails, printing and logging "mkdir" error
+ DietPi-Sync | Exclude specific files/dirs only, if they are actually inside the source dir, and, correctly via relative path. Update info on exclude/include user file
+ DietPi-Sync | Do not allow to choose root "/" as source or target dir, point user towards DietPi-Backup. As well check for correct full path entry with leading slash.
+ DietPi-Sync | Do not check for existing dietpi_userdata which is not necessarily required
+ DietPi-Sync | Return to main menu, if directory selection was successful
+ DietPi-Sync | Removed separate "delete mode" and "cron choice" menu IDs. Those are too simple to require an own ID and we always want to go back to main menu afterwards.
+ DietPi-Sync | Turn $SYNC_DRY_RUN and samba related variables into local ones
+ DietPi-Sync | Minor coding and wording consistency
+ CHANGELOG | Entries

MichaIng committed:
v6.20

+ DietPi-Sync | Resolved an issue with wrong created rsync options string
+ DietPi-Sync | Minor coding and wording consistency

MichaIng committed:
v6.20

+ CHANGELOG | DietPi-Sync fix

MichaIng committed:
v6.20

+ DietPi-Sync | Resolve doubled dry-run setting, quirk due to separate dry-run option and free space check dry-run

MichaIng committed:
v6.20

+ DietPi-Sync | Remove compression option, since this is only relevant for remote transfer, which we do not support yet
+ DietPi-Sync | Only run real sync, if dry-run from space check was successful

MichaIng commented:
@Fourdee 
Exclusion of directory when sync to nested dir does not work. Tested on all distro, so no bug of special rsync version. No idea how to achieve the sync target to be excluded then:
```
root@VM-Jessie:~# l
total 28
-rw------- 1 root root 5036 Dec 21 21:50 .bash_history
-rw-r--r-- 1 root root  560 Sep 25 15:06 .bashrc
drwxr-xr-x 3 root root 4096 May 20  2018 .config
-rw------- 1 root root   18 Dec 23 20:34 .nano_history
-rw-r--r-- 1 root root  140 Nov 19  2007 .profile
drwxr-xr-x 2 root root 4096 Oct 24 22:22 .ssh
root@VM-Jessie:~# rsync -av --delete --delete-excluded --exclude=/root/backup /root/ /root/backup/
sending incremental file list
created directory /root/backup
./
.bash_history
.bashrc
.nano_history
.profile
.config/
.config/htop/
.config/htop/htoprc
.ssh/
.ssh/known_hosts

sent 7,373 bytes  received 188 bytes  15,122.00 bytes/sec
total size is 6,804  speedup is 0.90
root@VM-Jessie:~# rsync -av --delete --delete-excluded --exclude=/root/backup /root/ /root/backup/
sending incremental file list
./
backup/
backup/.bash_history
backup/.bashrc
backup/.nano_history
backup/.profile
backup/.config/
backup/.config/htop/
backup/.config/htop/htoprc
backup/.ssh/
backup/.ssh/known_hosts

sent 7,642 bytes  received 156 bytes  15,596.00 bytes/sec
total size is 13,608  speedup is 1.75
root@VM-Jessie:~# rsync -av --delete --delete-excluded --exclude=/root/backup /root/ /root/backup/
sending incremental file list
backup/
backup/backup/
backup/backup/.bash_history
backup/backup/.bashrc
backup/backup/.nano_history
backup/backup/.profile
backup/backup/.config/
backup/backup/.config/htop/
backup/backup/.config/htop/htoprc
backup/backup/.ssh/
backup/backup/.ssh/known_hosts

sent 7,911 bytes  received 156 bytes  16,134.00 bytes/sec
total size is 20,412  speedup is 2.53
```

Actually the the reason is, that include/exclude does not take full file patchs, but only file name patterns... That breaks the whole exclusions:
```
root@VM-Jessie:~# rm -R backup/
root@VM-Jessie:~# rsync -av --delete --delete-excluded --exclude=/root/.profile /root/ /root/backup/
sending incremental file list
created directory /root/backup
./
.bash_history
.bashrc
.nano_history
.profile
.config/
.config/htop/
.config/htop/htoprc
.ssh/
.ssh/known_hosts

sent 7,373 bytes  received 188 bytes  15,122.00 bytes/sec
total size is 6,804  speedup is 0.90
root@VM-Jessie:~# rsync -av --delete --delete-excluded --exclude=.profile /root/ /root/backup/
sending incremental file list
deleting .profile
./
backup/
backup/.bash_history
backup/.bashrc
backup/.nano_history
backup/.config/
backup/.config/htop/
backup/.config/htop/htoprc
backup/.ssh/
backup/.ssh/known_hosts

sent 7,425 bytes  received 149 bytes  15,148.00 bytes/sec
total size is 13,328  speedup is 1.76
```
- Reading: https://stackoverflow.com/questions/5343558/cannot-get-rsync-exclude-option-to-exclude-dir

Damn, can't we exclude by full file patchs instead of relative ones only??

🈯️ The following works:
```
root@VM-Jessie:~# rsync -av --delete --delete-excluded --exclude=$(realpath --relative-to=/root /root/backup) /root/ /root/backup/
sending incremental file list
created directory /root/backup
./
.bash_history
.bashrc
.nano_history
.profile
.config/
.config/htop/
.config/htop/htoprc
.ssh/
.ssh/known_hosts

sent 7,369 bytes  received 184 bytes  15,106.00 bytes/sec
total size is 6,804  speedup is 0.90
root@VM-Jessie:~# rsync -av --delete --delete-excluded --exclude=$(realpath --relative-to=/root /root/backup) /root/ /root/backup/
sending incremental file list
./

sent 290 bytes  received 22 bytes  624.00 bytes/sec
total size is 6,804  speedup is 21.81
```

So all include/exclude entries need to be created like: `$(realpath --relative-to=$FP_SOURCE /full/path)`
- The question is how we handle this for end user. Read the custom include/exclude file and converting each line?

And does `--exclude-from` actually allow to include locations as well, since there is a separate `--include-from` option?
- Seems to, only `-` will be default then, so should be theoretically possible to add name patterns without leading `-`.

German ubuntu man page makes it very clear: https://wiki.ubuntuusers.de/rsync/#Verzeichnisse-von-der-Sicherung-ausschliessen

So: Add exclude pattern based on actual source dir (`if [[ /path/to/exclusion == $FP_SOURCE/* ]]`) and relative to it (`realpath --relative-to=$FP_SOURCE /path/to/exclusion >> $FP_FILTER_INCLUDE_EXCLUDE`).
- Add info to custom user inc_exc file that entries need to be patterns, in case with related path to source dir. Perhaps add link to man page.
__________
Other topic: Clear log file before new run + clear dry-run entries, before running real sync?

MichaIng committed:
v6.20

+ DietPi-Sync | Typo

Fourdee reviewed (approved):
@MichaIng 

Great work, looking good 👍 

Please merge when ready.

MichaIng committed:
v6.20

+ DietPi-Cron.hourly | Failsafe: Allow missing/invalid INDEX_LOGGING_CURRENT entry

MichaIng committed:
v6.20

+ DietPi-Cron.daily | Fix daily dietpi-sync based on new settings file scheme

MichaIng committed:
Merge branch 'dev' into dietpi-sync

MichaIng committed:
v6.20

+ DietPi-Patch | Remove obsolete compression entry from DietPi-Sync settings patch

MichaIng committed:
v6.20

+ DietPi-Sync | Exclude specific files/dirs only, if they are actually inside the source dir, and, correctly via relative path. Update info on exclude/include user file
+ DietPi-Sync | Do not allow to choose root "/" as source or target location, point user towards DietPi-Backup. As well check for correct full path entry with leading slash.
+ DietPi-Sync | Do not check for existing dietpi_userdata, since this is not necessarily required
+ DietPi-Sync | Syntax within free space check: Replace "break" by "return", since we are not in a loop

MichaIng committed:
v6.20

+ DietPi-Sync | Minor failsafe

MichaIng committed:
v6.20

+ DietPi-Sync | Typo
+ DietPi-Sync | Syntax: "declare" requires "-g", when called within a function, otherwise declarations are local only
+ DietPi-Sync | Return to main manu, if source/target directory has been set successfully

MichaIng committed:
v6.20

+ DietPi-Sync | Fix recent sync date text, not showing last status log, including timestamp and error message in case

MichaIng commented on dietpi/dietpi-sync:
```diff
@@ -140,79 +140,71 @@ _EOF_
 
 			#Rsync options
-			local rync_options+="-aHP4 --info=name0 --info=progress2 --human-readable --delete-excluded --exclude-from=$FP_FILTER_INCLUDE_EXCLUDE --log-file=$FP_LOG "
+			local rync_options="-aHP4h --info=name0 --info=progress2 --delete-excluded --exclude-from=$FP_FILTER_INCLUDE_EXCLUDE --log-file=$FP_LOG"
```
@Fourdee 
Due to `--delete-excluded`, the sync status file is deleted as well, so only the last status log will be present. Not a big issue for me, but due to grep/tail combi to check last status, I guess this was not intended?

The /var/log/dietpi-sync.log on the other hand will grow, if not dietpi-ramlog in use. Since this contains many lines on large sync dirs, perhaps we should clear it at sync start?

Or actually for my impression we could merge both into a single log file. To preserve status (with ramlog) e.g. log everything to `.dietpi-sync.log` (tmp working dir), incl. final status output, then `mv .dietpi-sync.log $FP_TARGET/`?
_______
Lol typo: `rync_options` instead of `rsync_options` 😄

Fourdee replied:
@MichaIng 

> Or actually for my impression we could merge both into a single log file. To preserve status (with ramlog) e.g. log everything to .dietpi-sync.log (tmp working dir), incl. final status output, then mv .dietpi-sync.log $FP_TARGET/?

Yep good idea 👍.

MichaIng committed:
v6.20

+ DietPi-Sync | Switch from sync options "Enable"/"Disable" wording  to "[On]"/"[Off]", to as well fix default values

MichaIng committed:
v6.20

+ DietPi-Sync | Simplify logging: Log to FP_TARGET only, create tmp log first, move to $FP_TARGET afterwards; Show final whip and log file only, if not known issue (e.g. insufficient space check).
+ DietPi-Sync | Remove delete mode and cron choice menus from separate menu ID. Those are too simple to require an own ID and we always want to go back to main menu afterwards

MichaIng committed:
v6.20

+ DietPi-Sync | Be more verbose when target dir pre-creation fails, printing and logging "mkdir" error
+ DietPi-Sync | Turn $SYNC_DRY_RUN and samba related variables into local ones

MichaIng committed:
v6.20

+ DietPi-Sync | Tiny coding
